pkg/plugins/manifest.go

```diff
@@ -51,9 +27,61 @@ type pluginManifest struct {
 	Files   map[string]string `json:"files"`
 }
 
+// ManifestKeys is the database representation of public keys
+// used to verify plugin manifests.
+type ManifestKeys struct {
+	ID        int64  `xorm:"id autoincr"`
+	KeyID     string `xorm:"key_id"`
+	PublicKey string
+	Since     int64
+	RevokedAt int64
+	UpdatedAt int64
+}
+
+type manifestVerifier struct {
+	sqlstore   *sqlstore.SqlStore
+	lock       sync.Mutex
+	publicKeys map[string]ManifestKeys
+}
+
+func newManifestVerifier(sqlstore *sqlstore.SqlStore) *manifestVerifier {
+	return &manifestVerifier{sqlstore: sqlstore}
+}
+
+// getPublicKey loads public keys from the database.
+// Soon we can updated keys from https://grafana.com/api/plugins/ci/keys
+func (pmv *manifestVerifier) getPublicKey(keyID string) (string, error) {
+	pmv.lock.Lock()
+	defer pmv.lock.Unlock()
+
+	// since we don't have any public keys we need to load them first
+	if len(pmv.publicKeys) == 0 {
+		session := pmv.sqlstore.NewSession()
+		defer session.Close()
+
+		var keys []ManifestKeys
+		err := session.Find(&keys)
+		if err != nil {
+			return "", err
+		}
+
+		pmv.publicKeys = make(map[string]ManifestKeys, len(keys))
+		for _, k := range keys {
+			pmv.publicKeys[k.KeyID] = k
+		}
+	}
+
+	key, exist := pmv.publicKeys[keyID]
+	if exist {
+		return key.PublicKey, nil
+	}
+
+	return "", errors.New("Could not find public Key")
+}
+
 // readPluginManifest attempts to read and verify the plugin manifest
 // if any error occurs or the manifest is not valid, this will return an error
-func readPluginManifest(body []byte) (*pluginManifest, error) {
+func (pmv *manifestVerifier) readPluginManifest(body []byte) (*pluginManifest, error) {
 	block, _ := clearsign.Decode(body)
```

Comment:
Not related to this PR I know, but figure we might as well fix it :)
```suggestion
	block, err := clearsign.Decode(body)
	if err != nil {
	        return nil, err
	}
```

Reply:
This one tripped me as well. But the second return is not an error. 
https://godoc.org/golang.org/x/crypto/openpgp/clearsign#Decode

Reply:
@bergquist Oops, my bad.